### DISCUSSION TITLE
thanhle.blog/blog/server-side-rendering-voi-hydration-lang-phi-tai-nguyen-nhu-the-nao

### DISCUSSION THREAD

**Discussion by giscus** [thanhlmm/blog#38] [Ideas]
# thanhle.blog/blog/server-side-rendering-voi-hydration-lang-phi-tai-nguyen-nhu-the-nao

Tìm hiểu về cách các app React, Vue, Svelte đang làm với Sever side rendering và nó lãng phí tài nguyên như thế nào

https://thanhle.blog/blog/server-side-rendering-voi-hydration-lang-phi-tai-nguyen-nhu-the-nao

**Comment by igdev116** [comment]
Bài viết hay lắm b phải làm mk login vào để lai. Nhưng có đoạn Website đc gửi xuống 2 lần có nghĩa là lần đầu server nó trả cục HTML cho client để render, sau đó server nó tiếp tục gọi lần 2 để trả về JS cho client để interactive. Mình hiểu vậy ko biết có đúng ko nữa 😂
```
Nói ngắn gọn, hydration là công việc recover lại những hàm xử lý event bằng cách Download code, chạy lại các component mà SSR/SSG gửi xuống. Do đo website sẽ được gửi xuống 2 lần, 1 là dưới dạng HTML, 2 là dưới dạng JavaScript.
```

**Reply by thanhlmm** [reply]
Mình giả sử bạn có một component A như vầy nhé
```jsx

const A = () => {
  const sayGreeting = () => alert("Bonjour");
  
  return <div onClick={sayGreeting}>Hello world</div>
}
```

Thì khi gửi xuống browser HTML sẽ là (Render lần 1 trên server)
```html
 <div>Hello world</div>
```
Đồng thời cũng sẽ kèm theo một đoạn code js tương đương như component trên để phục vụ quá trình Hydration. JS sau khi được download và chạy lại (Render lần 2 dưới client) để gắn `onClick={sayGreeting}` vào cây DOM

Vậy là code của mình chạy 2 lần để render 1 component đó
